### PR TITLE
container parallel testing

### PR DESCRIPTION
**Pull request by ansvu**
build-depends: 31590
build-depends: https://github.com/dci-labs/bos2-ci-config/pull/148

Status since last commit 09/Aug/2024:
---

- [x] preflight-green https://www.distributed-ci.io/jobs/6c57a3fd-639d-4a5c-876b-db52671b3114/jobStates?sort=date
- [x] control-plane-connect-parallel https://www.distributed-ci.io/jobs/8c854e13-17a6-4bee-97ad-39f3a588324a/jobStates?sort=date
- [x] control-plane-15-parallel https://www.distributed-ci.io/jobs/c57b22b9-1b5f-4b67-8f3b-cfe2f5e1c7a5/jobStates?sort=date
- [x] control-plane-no-parallel https://www.distributed-ci.io/jobs/b31867ba-3be6-4aa7-a505-da0a9137987e/jobStates?sort=date
- [x] control-plane-parallel-projects-public https://www.distributed-ci.io/jobs/1528b4ca-2f10-40ca-91b4-ce63fba0c36d/jobStates?sort=date
- [x] control-plane-parallel-projects-public-other https://www.distributed-ci.io/jobs/05dcc77d-14c8-4cb4-8320-27913f2411a6/jobStates?sort=date
- [x] control-plane-connect-parallel -https://www.distributed-ci.io/jobs/4333cf1f-b7d6-444f-80f4-f1a5de32a0c6/jobStates?sort=date (mixed scenario new and existing projects)

Test-Hints: no-check

### CODE REVIEW

File: roles/preflight/defaults/main.yml [hunk 1]
```diff
@@ -8,4 +8,6 @@ preflight_podman_ca: ""
 catalog_url: "https://catalog.redhat.com/api/containers/v1"
 preflight_test_certified_image: false
 preflight_run_health_check: true
+do_container_parallel_test: false
```
[the remark this answers is not in the document]
Comment (ansvu) [reply roles/preflight/defaults/main.yml]: Ack. Thanks.

File: roles/preflight/README.md [hunk 2]
```diff
@@ -20,6 +20,8 @@ preflight_test_certified_image  | false
 catalog_url                       | https://catalog.redhat.com/api/containers/v1         | Optional. This is a Pyxis API that used during the check if the image is certified.
 preflight_run_health_check      | true                                                 | Optional. Run health check on every container and generate oval reports both in xml and HTML formats.
 preflight_dci_all_components_are_ga | true                                             | Optional. Only submit test results when all components in the list `dci_ga_components_for_certification` are GA.
+do_container_parallel_test | false                                             | Optional. Run Preflight test in parallel for container that need to recertify or without submission to the portal.
```
Comment (ramperher) [inline roles/preflight/README.md:23]: This entry can also be removed then :)

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
retest

File: roles/preflight/tasks/main_standalone_containers_certification.yml [hunk 3]
```diff
@@ -34,6 +34,18 @@
       loop: "{{ preflight_containers_to_certify }}"
       loop_control:
         loop_var: c
+      when:
+        - not do_container_parallel_test
+        - max_images_per_batch == 1
```
Comment (thekad) [inline roles/preflight/tasks/main_standalone_containers_certification.yml:39]: there's a small chance someone passes -1 here and none of the checks run, change this to `max_images_per_batch <= 1`

File: roles/preflight/README.md (hunk 2 above)
Comment (ramperher) [reply roles/preflight/README.md]: resolved

File: roles/preflight/tasks/main_standalone_containers_certification.yml (hunk 3 above)
Comment (ansvu) [reply roles/preflight/tasks/main_standalone_containers_certification.yml]: Thanks, Jorge!

**Comment by manurodriguez** [comment]
Status since last commit:
---

- [x] preflight-green- https://www.distributed-ci.io/jobs/f79beec3-1955-4a18-b7db-a5b9ac4a853b/jobStates
- [x] control-plane no parallel - https://www.distributed-ci.io/jobs/a24f2c48-1532-4a21-92b8-2a89adc4e5b2/jobStates?sort=date
- [x] control-plane 1 parellel - https://www.distributed-ci.io/jobs/4c9fe71a-b0ee-4983-9964-d554fac505e4/jobStates?sort=date
- [x] all feedback and comments resolved

---
overall looks good to me, I talked with Ying and Andrew to wait a little bit for @tkrishtop review, thanks!

File: roles/preflight/tasks/main_standalone_containers_certification.yml [hunk 4]
```diff
@@ -34,6 +34,18 @@
       loop: "{{ preflight_containers_to_certify }}"
       loop_control:
         loop_var: c
+      when:
+        - not do_container_parallel_test
```
Comment (ramperher) [inline roles/preflight/tasks/main_standalone_containers_certification.yml:38]: I've seen this variable is declared in dci-openshift-app-agent, but not used there. What happens if anyone tries to use this role, just launching it from Ansible and without DCI? Then, we should provide a default value here, and don't know if it's worth to add it to the README too.

**Review by ramperher — COMMENTED** [review]
Now that I've seen the final code, just one minor comment from my side.

**Review by tkrishtop — CHANGES_REQUESTED** [review]
I tested the standard flow "run preflight tests + create certification projects", here is the [job](https://www.distributed-ci.io/jobs/6e948e3d-c2cd-4dea-b804-cf73a7f8d266/jobStates?sort=date).

```
    preflight_containers_to_certify:
      - container_image: "quay.io/avu0/demo-parallel-x1:v1"
        create_container_project: true
      - container_image: "quay.io/avu0/demo-parallel-x2:v1"
        create_container_project: true
```

The creation of Connect projects is simply ignored when using parallelization flow. We have to add it the same way the creation of tmp folders is currently done.

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by manurodriguez** [comment]
recheck

**Comment by manurodriguez** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by manurodriguez** [comment]
Testing to close and re-open to see if that triggers the CI job

**Comment by manurodriguez** [comment]
recheck

**Comment by manurodriguez** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
retest

**Comment by ansvu** [comment]
recheck

**Comment by ansvu** [comment]
retest

**Comment by tkrishtop** [comment]
hi @ansvu no worries about the force-push, I simply rebased your changes over the main branch to unblock the testing

**Comment by manurodriguez** [comment]
retest

**Comment by manurodriguez** [comment]
recheck

**Comment by manurodriguez** [comment]
recheck

**Comment by manurodriguez** [comment]
retest

**Comment by manurodriguez** [comment]
recheck

**Review by tkrishtop — APPROVED** [review]
thank you for your hard work @ansvu !

**Comment by manurodriguez** [comment]
recheck

**Pull request merged**